src/metabase/driver/sql_jdbc/execute.clj

```diff
@@ -332,12 +338,16 @@
   Connection."
   {:added "0.47.0"}
   [driver                                                 :- :keyword
+   db-or-id-or-spec
    ^Connection conn                                       :- (lib.schema.literal.jvm/instance-of Connection)
    {:keys [^String session-timezone write?], :as options} :- ConnectionOptions]
   (when-not (recursive-connection?)
     (log/tracef "Setting default connection options with options %s" (pr-str options))
     (set-best-transaction-level! driver conn)
     (set-time-zone-if-supported! driver conn session-timezone)
+    (set-role-if-supported! driver conn (when (u/id db-or-id-or-spec) (if (integer? db-or-id-or-spec)
+                                                                        (t2/select-one Database db-or-id-or-spec)
+                                                                        db-or-id-or-spec)))
```

Comment:
We need to fetch the database here for `set-role-if-supported!` since the default role can be configured per-DB for Snowflake. Unfortunately if `db-or-id-or-spec` is already a spec, we can't fetch the DB, so conn impersonation would fall back to `ACCOUNTADMIN`. Not sure if there's a better way to do this without removing the ability to pass in just a spec here.